### PR TITLE
[java] Update ArrayIsStoredDirectly and MethodReturnsInternalArray

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Update ArrayIsStoredDirectly and MethodReturnsInternalArray, remove their abstract base class.

ArrayIsStoredDirectly now reports on the assignment and not the formal parameter so line numbers will probably move. May be worth mentioning in #3123 

## Related issues

- Items of #2701 
- Related to #971 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 466 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/652df96d20d44f129cb64bf90f3b36bd/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 2334 violations,<br />
introduces 6084 new violations, 2 new errors and 0 new configuration errors,<br />
removes 15193 violations, 10 errors and 2 configuration errors.<br />
<a href="https://chunk.io/pmd/652df96d20d44f129cb64bf90f3b36bd/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by oowekyala** [comment]
Summary of the differences,
* MethodReturnsInternalArray does not report on static fields anymore -> it probably needs to
  * https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-core/src/test/java/org/springframework/core/PrioritizedParameterNameDiscovererTests.java#L37 
* MethodReturnsInternalArray now reports on `this.field` -> this is expected
  * https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-jdbc/src/main/java/org/springframework/jdbc/support/CustomSQLErrorCodesTranslation.java#L51
* ArrayIsStoredDirectly violations moved
* Only [one violation](https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-aop/src/main/java/org/springframework/aop/framework/ReflectiveMethodInvocation.java#L239) is removed for ArrayIsStoredDirectly. I think this is expected as the field is set on another object as `this`. Otherwise the rule should consider all field stores, which it doesn't.

**Comment by adangel** [comment]
Yes, ArrayIsStoredDirectly looks fine.

For MethodReturnsInternalArray, I see these two issues:
* false negative: static fields - as you described. Yes, we should report those as well
* false positive: there is one here: https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-jdbc/src/main/java/org/springframework/jdbc/core/JdbcTemplate.java#L574
The method here just returns "String" and no array. But it is declared in a class, which is declared inside a method, which returns an array (https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-jdbc/src/main/java/org/springframework/jdbc/core/JdbcTemplate.java#L518) - we probably got confused by that.

I'll try to look into these two issues now....

**Comment by adangel** [comment]
@oowekyala I've pushed some commits:
* The two cases are covered. The static field case is a bit special - if the field is declared final, it is considered "constant" and it is ok to return it (see https://sourceforge.net/p/pmd/bugs/1475/). That's of course debatable, because the array is indeed modifiable. But that's how the rule works on pmd6.
* The specific case (https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-core/src/test/java/org/springframework/core/PrioritizedParameterNameDiscovererTests.java#L37) is actually a false-positive on pmd6 - since the field is final.
* I'm not sure, whether the findBoundries is correct on ASTLocalClassStatement. I've tried to put it e.g. in ASTClassOrInterfaceDeclaration, but that seemed to break more things, so I reverted it again (https://github.com/pmd/pmd/pull/3127/commits/a1d36106a596a734597289f45fd26d0cd28bea05 https://github.com/pmd/pmd/pull/3127/commits/9bee4564c7541d8e808b14114826a20ef0df51ab)

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/MethodReturnsInternalArray.xml [hunk 1]
```diff
@@ -366,6 +366,52 @@ public class OuterClass {
...
+        <expected-linenumbers>16</expected-linenumbers>
+        <code><![CDATA[
+import java.util.concurrent.Callable;
+
+public class MyClass {
+    private static final String[] FOO_BAR = new String[] { "foo", "bar" };
+    private final Callable<String[]> returnsFooBar = new Callable<String[]>() {
+        @Override
+        public String[] call() {
+            return FOO_BAR; // no violation, because FOO_BAR is final. See https://sourceforge.net/p/pmd/bugs/1475/
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/MethodReturnsInternalArray.xml:404]: In fact this is only safe is the array has length zero... right? We could check the initializer for `{}` and `new X[0]`.

Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/MethodReturnsInternalArray.xml:404]: That's correct, I guess. I'm hesitating a bit, fixing this bug with this PR. Maybe we should create an issue for this an fix in separately? I think, the bug applies also for pmd6 and fix for https://sourceforge.net/p/pmd/bugs/1475/ needs to be reworked...

Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/MethodReturnsInternalArray.xml:404]: I think you're right that this needs a ticket to track the issue on master, but it's probably simpler to fix it in this PR for pmd 7 (as a fix on master would need to be adapted to te

Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/MethodReturnsInternalArray.xml:404]: -> #3160

**Review by adangel — APPROVED** [review]

**Pull request merged**